File: app/src/main/java/com/serwylo/lexica/db/GameMode.java
package com.serwylo.lexica.db;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;

@AllArgsConstructor
@Entity
@Data
@Builder
public class GameMode implements Parcelable {

    public static final String HINT_MODE = "hintMode";
    public static final String SCORE_TYPE = "scoreType";
    public static final String SCORE_WORDS = "W";
    public static final String SCORE_LETTERS = "L";

    @PrimaryKey(autoGenerate = true)
    private long gameModeId;

    @NonNull
    private String label;

    @NonNull
    private String description;

    /**
     * 16, 25, 36.
     */
    private int boardSize;

    private int timeLimitSeconds;

    /**
     * Between 3 and 9.
     */
    private int minWordLength;

    @NonNull
    private String scoreType;

    @NonNull
    private String hintMode;

    private boolean isCustom;

    public GameMode(@NonNull String label, @NonNull String description, int boardSize, int timeLimitSeconds, int minWordLength, @NonNull String scoreType, @NonNull String hintMode, boolean isCustom) {
        this.label = label;
        this.description = description;
        this.boardSize = boardSize;
        this.timeLimitSeconds = timeLimitSeconds;
        this.minWordLength = minWordLength;
        this.scoreType = scoreType;
        this.hintMode = hintMode;
        this.isCustom = isCustom;
    }

    protected GameMode(Parcel in) {
        gameModeId = in.readLong();
        label = in.readString();
        description = in.readString();
        boardSize = in.readInt();
        timeLimitSeconds = in.readInt();
        minWordLength = in.readInt();
        scoreType = in.readString();
        hintMode = in.readString();
        isCustom = in.readInt() == 1;
    }

    public boolean hintModeCount() {
        return hintMode.equals("tile_count") || hintMode.equals("hint_both");
    }

    public boolean hintModeColor() {
        return hintMode.equals("hint_colour") || hintMode.equals("hint_both");
    }

    public static final Creator<GameMode> CREATOR = new Creator<GameMode>() {
        @Override
        public GameMode createFromParcel(Parcel in) {
            return new GameMode(in);
        }

        @Override
        public GameMode[] newArray(int size) {
            return new GameMode[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(gameModeId);
        dest.writeString(label);
        dest.writeString(description);
        dest.writeInt(boardSize);
        dest.writeInt(timeLimitSeconds);
        dest.writeInt(minWordLength);
        dest.writeString(scoreType);
        dest.writeString(hintMode);
        dest.writeInt(isCustom ? 1 : 0);
    }
}

File: app/src/main/java/com/serwylo/lexica/db/GameMode.kt
package com.serwylo.lexica.db

import android.content.Context
import android.os.Parcelable
import androidx.room.PrimaryKey
import androidx.room.Entity
import com.serwylo.lexica.R
import kotlinx.android.parcel.Parcelize
import kotlin.math.sqrt


@Parcelize
@Entity
data class GameMode (

        @PrimaryKey(autoGenerate = true)
        var gameModeId: Long = 0,

        val type: Type,

        val customLabel: String? = null,

        /**
         * 16, 25, 36.
         */
        val boardSize: Int,

        val timeLimitSeconds: Int,

        /**
         * Between 3 and 9.
         */
        val minWordLength: Int,

        val scoreType: String,

        val hintMode: String,

) : Parcelable {

    enum class Type {
        SPRINT,
        MARATHON,
        BEGINNER,
        LETTER_POINTS,
        CUSTOM,
        LEGACY,
    }

    fun label(context: Context): String {
        return when (type) {
            Type.SPRINT -> context.getString(R.string.game_mode_sprint)
            Type.MARATHON -> context.getString(R.string.game_mode_marathon)
            Type.BEGINNER -> context.getString(R.string.game_mode_beginner)
            Type.LETTER_POINTS -> context.getString(R.string.game_mode_letter_points)

            Type.LEGACY -> {
                val duration = context.resources.getQuantityString(R.plurals.num_minutes, timeLimitSeconds / 60, timeLimitSeconds / 60)
                val boardWidth = sqrt(boardSize.toDouble()).toInt()
                val size = "${boardWidth}x${boardWidth}"
                val score = if (SCORE_LETTERS == scoreType) context.getString(R.string.letter_points) else context.getString(R.string.word_length)
                return "$duration / $size / $score"
            }

            // Don't bother internationalizing, because there should always be a custom label
            // when the type is custom, though Kotlin doesn't know this.
            Type.CUSTOM -> customLabel ?: "Custom"
        }
    }

    fun description(context: Context): String {
        return when (type) {
            Type.SPRINT -> context.getString(R.string.game_mode_sprint_description)
            Type.MARATHON -> context.getString(R.string.game_mode_marathon_description)
            Type.BEGINNER -> context.getString(R.string.game_mode_beginner_description)
            Type.LETTER_POINTS -> context.getString(R.string.game_mode_letter_points_description)
            Type.CUSTOM -> context.getString(R.string.game_mode_custom_description)
            Type.LEGACY -> context.getString(R.string.game_mode_legacy_description)
        }
    }

    fun hintModeCount(): Boolean {
        return hintMode == "tile_count" || hintMode == "hint_both"
    }

    fun hintModeColor(): Boolean {
        return hintMode == "hint_colour" || hintMode == "hint_both"
    }

    fun serialize(): String {
        return StringBuilder()
                .appendLine(gameModeId)
                .appendLine(type.name)
                .appendLine(customLabel ?: "")
                .appendLine(boardSize)
                .appendLine(timeLimitSeconds)
                .appendLine(minWordLength)
                .appendLine(scoreType)
                .appendLine(hintMode)
                .toString()
    }

    override fun toString(): String {
        return "Game Mode [id: $gameModeId, type: $type]"
    }

    companion object {
        const val SCORE_WORDS = "W"
        const val SCORE_LETTERS = "L"

        @JvmStatic fun deserialize(modeString: String): GameMode {
            val parts = modeString.split("\n")

            return GameMode(
                    gameModeId = parts[0].toLong(),
                    type = Type.valueOf(parts[1]),
                    customLabel = if (parts[2] == "") null else parts[2],
                    boardSize = parts[3].toInt(),
                    timeLimitSeconds = parts[4].toInt(),
                    minWordLength = parts[5].toInt(),
                    scoreType = parts[6],
                    hintMode = parts[7],
            )
        }

    }

}